### PR TITLE
feat: add test for the section challenge

### PR DESCRIPTION
**Pull request by Samantha-KY**
# Submit a pull request

- [ ] This is not a duplicate of an existing pull request.
- [ ] No existing features have been broken without good reason.
- [ ] Your commit messages are detailed
- [ ] The code style guideline have been followed.
- [ ] Documentation has been updated to reflect your changes.
- [ ] Tests have been added or updated to reflect your changes.
- [ ] All tests pass.

---

Replace any ":question:" below with information about your pull request.

## Pull Request Details

Provide details about your pull request and what it adds, fixes, or changes.

:question:

## Breaking Changes

Describe what features are broken by this pull request and why, if any.

:question:

## Issues Fixed

Enter the issue numbers resolved by this pull request below, if any.

1. :question:

## Other Relevant Information

Provide any other important details below.

### CODE REVIEW

File: __tests__/components/sections/challenges/Objectives.test.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+describe("Objectives", () => {
+  const challenge = mockChallenge();
+  it("should render objectives", () => {
+    renderWithRedux(<Objectives testId="objectiveId" />);
+    expect(screen.getByTestId("objectiveId")).toBeInTheDocument();
+    expect(screen.getByText("communities.overview.challenge.objective.title")).toBeInTheDocument();
+  });
+
+  it("should render the objective list", () => {
+    renderWithRedux(<Objectives />, { challenges: { current: mockChallenge(), list: [mockChallenge()], loading: false, submission: submission() } });
```
Comment (mhenriette) [inline __tests__/components/sections/challenges/Objectives.test.tsx:17]: can you create a mock object for the data and use it everywhere it is needed instead of doing this

File: src/components/sections/challenges/Rubric.tsx [hunk 2]
```diff
@@ -57,7 +58,7 @@ export default function RubricHeader({ ratingCriteria, selected, hideTitle = fal
     const hackatonPassingScore = t("communities.challenge.hackathon.passing.score", { threshold: challenge?.threshold, prizePool: `USD ${reward?.amount}` });
 
     return (
-      <div className="text-base font-normal text-slate-700 pt-6 inline-flex flex-wrap items-center gap-1">
+      <div data-testid={testId} className="text-base font-normal text-slate-700 pt-8 md:w-182.5 inline-flex flex-wrap items-center gap-1">
```
Comment (mhenriette) [inline src/components/sections/challenges/Rubric.tsx:61]: why add padding and width sizes ?

File: src/components/sections/challenges/SetupTeamChallenge.tsx [hunk 3]
```diff
@@ -18,15 +18,15 @@ interface Props {
...
+export default function SetupTeamChallenge({testid}: {testid?: string}): JSX.Element {
   const { invite, challenge } = useMultiSelector<unknown, Props>({ invite: (state: IRootState) => state.invites.data, challenge: (state: IRootState) => state.challenges.current });
 
   const { t } = useTranslation();
 
   return (
     <Section title="Submission">
-      <div className="text-base font-normal text-slate-700 py-6 md:w-182.5">{t("communities.overview.challenge.team.setup.info")}</div>
-      <div className="flex flex-col md:flex-row gap-6">
+      <div data-testid={testid} className="text-base font-normal text-slate-700 pt-8 pb-7 md:w-182.5">{t("communities.overview.challenge.team.setup.info")}</div>
```
Comment (mhenriette) [inline src/components/sections/challenges/SetupTeamChallenge.tsx:28]: same here

File: __tests__/components/sections/challenges/TeamChallenge.test.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      text: "Make sure your teammates accept notification to confirm your team",
+    },
+    {
+      index: 3,
+      title: "Submit!",
+      text: "Once you have completed the challenge, only one person needs to submit it at the end of this page",
+    },
+  ];
+  it("should render the team challenge section", () => {
+    renderWithRedux(<TeamChallenge />, { challenges: { current: challenge(), list: [challenge()], loading: false, submission: submission() } });
```
Comment (mhenriette) [inline __tests__/components/sections/challenges/TeamChallenge.test.tsx:49]: create a mock variable for these data

File: __tests__/components/sections/challenges/SetupTeamChallenge.test.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  it("renders content correctly", () => {
+    renderWithRedux(<SetupTeamChallenge />, { challenges: { current: challenge, list: [challenge], loading: false, submission: submission() }, invites: { data: mockInvite } });
+    expect(screen.getByText("Submission")).toBeInTheDocument();
+    expect(screen.getByText("communities.overview.challenge.team.setup.info")).toBeInTheDocument();
+    expect(screen.getByText("Form your team")).toBeInTheDocument();
+    expect(screen.getByText(challenge.additionalInfo.TEAM_FORMATION.text) || "communities.overview.challenge.team.organization").toBeInTheDocument();
+  });
+  
+  it("renders CreateTeamCard when there is no invitation or comfirmTeamInvitation when there is invitation", () => {
+    renderWithRedux(<SetupTeamChallenge />, { challenges: { current: challenge, list: [challenge], loading: false, submission: submission() }, invites: { data: mockInvite } });
```
Comment (mhenriette) [inline __tests__/components/sections/challenges/SetupTeamChallenge.test.tsx:23]: same here, create  a mock variable

**Review by mhenriette — CHANGES_REQUESTED** [review]

File: __tests__/components/sections/challenges/Learning.test.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  });
+
+  it("should render with courses", () => {
+    const coursesArray = [mockCourse];
+    renderWithRedux(<Learning courses={[mockCourse]} learningModules={[mockLearningModule]} community={mockCommunity} />, {
+      challenges: { current: challenge(), list: [challenge()], loading: false, submission: submission() },
+    });
+    coursesArray.forEach((course) => {
+      expect(screen.getByText(course.name)).toBeInTheDocument();
+      expect(screen.getByText(course.description)).toBeInTheDocument();
```
Comment (mhenriette) [inline __tests__/components/sections/challenges/Learning.test.tsx:24]:
```suggestion
 const courseNameElement = screen.getByText(course.name);
      const courseDescriptionElement  = screen.getByText(course.description)
      expect(courseNameElement).toBeInTheDocument();
      expect(courseNameElement).toHaveTextContent(course.name);
      expect(courseDescriptionElement).toBeInTheDocument();
      expect(courseDescriptionElement).toHaveTextContent(course.description);
```

File: __tests__/components/sections/challenges/Learning.test.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    });
+  });
+
+  it("should render with learning", () => {
+    const learningArray = [mockLearningModule];
+    renderWithRedux(<Learning courses={[mockCourse]} learningModules={[mockLearningModule]} community={mockCommunity} />);
+    learningArray.forEach((learningModule) => {
+      expect(screen.getByText(learningModule.title)).toBeInTheDocument();
+      expect(screen.getByText(learningModule.description)).toBeInTheDocument();
```
Comment (mhenriette) [inline __tests__/components/sections/challenges/Learning.test.tsx:33]: same here, we should test the expected value

File: __tests__/components/sections/challenges/Objectives.test.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    renderWithRedux(<Objectives />, mockObjectivesStates);
+    const objectiveValue = challenge.objectives.find((objective) => objective);
+    expect(screen.getByText(objectiveValue || "objectives 1")).toBeInTheDocument();
+  });
+
+  it("should display expiry date", () => {
+    renderWithRedux(<Objectives />, mockObjectivesStates);
+    const expirationDate = challenge.expiresAt && DateManager.format(challenge.expiresAt, "MMMM d, yyyy", "en");
+    if (expirationDate) {
+      expect(screen.getByText(expirationDate)).toBeInTheDocument();
```
Comment (mhenriette) [inline __tests__/components/sections/challenges/Objectives.test.tsx:28]: same here

File: src/components/sections/challenges/Learning.tsx [hunk 9]
```diff
@@ -24,8 +24,8 @@ export default function Learning({ courses, learningModules, community }: { cour
         isExpanded
         content={
           <>
-            <div className="text-base font-normal text-slate-700 py-6">{t("communities.overview.challenge.learning.title")}</div>
-            <div className="grid grid-cols-1 md:grid-cols-2 gap-3 mb-3">
+            <div data-testid={testId} className="text-base font-normal text-slate-700 pt-8 pb-7 md:w-182.5">{t("communities.overview.challenge.learning.title")}</div>
```
Comment (mhenriette) [inline src/components/sections/challenges/Learning.tsx:27]: do we need these classes in testing ?

File: src/components/sections/challenges/Learning.tsx [hunk 10]
```diff
@@ -24,8 +24,8 @@ export default function Learning({ courses, learningModules, community }: { cour
...
+            <div className="md:grid grid-cols-2 gap-3">
```
Comment (mhenriette) [inline src/components/sections/challenges/Learning.tsx:28]: same as this breakpoint

File: src/components/sections/challenges/Rewards.tsx [hunk 11]
```diff
@@ -13,15 +13,15 @@ import RewardCertificate from "@/components/cards/challenge/RewardCertificate";
...
+export function OverviewRewards({testId}: {testId?: string}): ReactElement {
   const { t } = useTranslation();
   const challenge = useSelector((state) => state.challenges.current);
   const router = useRouter();
   const token = challenge?.reward?.token || challenge?.rewards[0]?.token || "";
 
   return (
     <Section title={`${t("communities.overview.reward.title")}`}>
-      <p className="my-6 text-lg">{t("course.challenge.reward.certificate.description")}</p>
+      <p data-testid={testId} className="my-5 text-lg">{t("course.challenge.reward.certificate.description")}</p>
```
Comment (mhenriette) [inline src/components/sections/challenges/Rewards.tsx:24]: same here , the margin, why do we need it ?

**Review by mhenriette — CHANGES_REQUESTED** [review]

**Review by mbukeRepo — COMMENTED** [review]
Avoid Hardcoding Data for Testing; Use Faker.js Instead

For more robust and flexible testing, it is recommended to avoid hardcoding data. Instead, consider using Faker.js. This library allows you to generate realistic and varied test data dynamically, which can enhance the reliability and thoroughness of your testing processes. By utilizing Faker.js, you can simulate a wide range of scenarios and edge cases, ensuring that your code is well-tested and resilient to different inputs. This practice not only improves test coverage but also enhances the maintainability of your test suite.

https://www.npmjs.com/package/@faker-js/faker

File: src/components/sections/challenges/Learning.tsx (hunk 10 above)
Comment (mhenriette) [reply src/components/sections/challenges/Learning.tsx:28]: you missed this review

File: __tests__/components/sections/challenges/Submission.test.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        expect(screen.getByText(errors.githubLink?.message as string)).toBeInTheDocument();
+      }
+      expect(screen.getByText("Markdown")).toBeInTheDocument();
+      if (challenge?.format.disclaimer) {
+        expect(screen.getByRole("checkbox")).toBeInTheDocument();
+      }
+      expect(screen.getByText("submit")).toBeInTheDocument();
+    }
+  });
+});
```
Comment (serapieTuyishime) [inline __tests__/components/sections/challenges/Submission.test.tsx:49]: The submission form is arleady tested

**Review by serapieTuyishime — CHANGES_REQUESTED** [review]

File: __tests__/components/sections/challenges/Rating.test.tsx
[the remark this answers is not in the document]
Comment (Samantha-KY) [reply __tests__/components/sections/challenges/Rating.test.tsx:1]: If you check well it is RubricRating, unless I am not getting what you mean

File: __tests__/components/sections/challenges/Rewards.test.tsx [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    expect(screen.getByTestId("overviewRewardId")).toBeInTheDocument();
+  });
+
+  it("should render reward with the challenge", () => {
+    renderWithRedux(<OverviewRewards testId="overviewRewardId" />, { challenges: { current: challenge, list: [challenge], loading: false, submission: submission } });
+    if (challenge.rewards) {
+      challenge.rewards.forEach((reward) => {
+        expect(screen.getByText(`${reward.amount} ${reward.token}`)).toBeInTheDocument();
+        expect(screen.getByText(`${reward.amount} ${reward.token}`).textContent).toBe("10 token");
+      });
```
[the remark this answers is not in the document]
Comment (Samantha-KY) [reply __tests__/components/sections/challenges/Rewards.test.tsx:25]: I removed the hardcoded values, but as the text to test contain both amount and token as one text the test will fail if I split them. unless I test if it contains one of the text?

File: __tests__/components/sections/challenges/Submission.test.tsx (hunk 12 above)
Comment (Samantha-KY) [reply __tests__/components/sections/challenges/Submission.test.tsx:49]: where was it tested?

**Review by serapieTuyishime — APPROVED** [review]

File: __tests__/components/sections/challenges/Learning.test.tsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+describe("Learning", () => {
+  it("should render Learning", () => {
+    renderWithRedux(<Learning testId="learningId" courses={[]} learningModules={[]} community={mockCommunity} />);
+    const learning = screen.getByTestId("learningId");
+    expect(learning).toBeInTheDocument();
+    expect(screen.getByText("communities.overview.challenge.learning.title")).toBeInTheDocument();
+  });
+
+  it("should render with course cards when they the challenge has them", () => {
```
Comment (Jonath-z) [inline __tests__/components/sections/challenges/Learning.test.tsx:17]:
```suggestion
  it("should render with course cards when the the challenge has them", () => {
```

File: __tests__/components/sections/challenges/SetupTeamChallenge.test.tsx [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+  it("renders content correctly", () => {
+    renderWithRedux(<SetupTeamChallenge />, mockTeamChallengeStates);
+    expect(screen.getByText("Submission")).toBeInTheDocument();
+    expect(screen.getByText("communities.overview.challenge.team.setup.info")).toBeInTheDocument();
+    expect(screen.getByText("Form your team")).toBeInTheDocument();
+    expect(screen.getByText(challenge?.additionalInfo?.TEAM_FORMATION.text || " ") || "communities.overview.challenge.team.organization").toBeInTheDocument();
+  });
+
+  it("renders CreateTeamCard when there is no invitation or comfirmTeamInvitation when there is invitation", () => {
```
Comment (Jonath-z) [inline __tests__/components/sections/challenges/SetupTeamChallenge.test.tsx:25]: This description is not clear enough

File: src/components/sections/challenges/Header.tsx [hunk 16]
```diff
@@ -11,12 +11,12 @@ import { ReactElement } from "react";
  * @export
  * @returns {ReactElement}
  */
-export default function ChallengeHeader(): ReactElement {
+export default function ChallengeHeader({testId}: {testId?: string}): ReactElement {
```
Comment (Jonath-z) [inline src/components/sections/challenges/Header.tsx:14]: testId should have a default value

File: src/components/sections/challenges/Learning.tsx [hunk 17]
```diff
@@ -13,7 +13,17 @@ import { useTranslation } from "next-i18next";
  *
  * @returns {JSX.Element} The Learning component JSX element.
  */
-export default function Learning({ courses, learningModules, community }: { courses: Course[]; learningModules: LearningModule[]; community: Community }): JSX.Element {
+export default function Learning({
+  courses,
+  learningModules,
+  community,
+  testId,
```
Comment (Jonath-z) [inline src/components/sections/challenges/Learning.tsx:20]: same here, apply give a default value to test id

File: src/components/sections/challenges/Rewards.tsx [hunk 18]
```diff
@@ -13,15 +13,17 @@ import RewardCertificate from "@/components/cards/challenge/RewardCertificate";
  * @export
  * @returns {ReactElement}
  */
-export function OverviewRewards(): ReactElement {
+export function OverviewRewards({ testId }: { testId?: string }): ReactElement {
```
Comment (Jonath-z) [inline src/components/sections/challenges/Rewards.tsx:16]: same here

File: src/components/sections/challenges/Rewards.tsx [hunk 19]
```diff
@@ -13,15 +13,17 @@ import RewardCertificate from "@/components/cards/challenge/RewardCertificate";
...
   const { t } = useTranslation();
   const challenge = useSelector((state) => state.challenges.current);
   const router = useRouter();
   const token = challenge?.reward?.token || challenge?.rewards[0]?.token || "";
 
   return (
     <Section title={`${t("communities.overview.reward.title")}`}>
-      <p className="my-6 text-base text-secondary">{t("course.challenge.reward.certificate.description")}</p>
+      <p data-testid={testId} className="my-6 text-base text-secondary">
```
Comment (mhenriette) [inline src/components/sections/challenges/Rewards.tsx:24]: why do you need this data-testId

**Review by mhenriette — CHANGES_REQUESTED** [review]

File: src/components/sections/challenges/Rewards.tsx (hunk 19 above)
Comment (Samantha-KY) [reply src/components/sections/challenges/Rewards.tsx:24]: to test the content inside the reward as the section component is already tested

Comment (mhenriette) [reply src/components/sections/challenges/Rewards.tsx:28]: can you use `getByText` to get the content instead of adding a `data-testid`
